### PR TITLE
Only request a single repository tag when pulling a specific image:tag

### PR DESCRIPTION
**Pull request by dkjer**
Currently when performing a docker pull, we call GetRemoteTags which pulls all tags for a repository, even though we are only interested in a single tag.  This request causes the pull to time out if there are a large number of tags on a given repository.

This PR changes GetRemoteTags to only pull the tag we are interested in, greatly reducing pull times in some scenarios.

### CODE REVIEW

**Comment by dmcgowan** [comment]
@dkjer is this intended to align the behavior of pulling v1 to that of v2, which only adds a tag for the specific tag requested? Originally this v2 behavior change was contested, but it was ultimately decided that it is better to only change tags based on what a user requests. I am assuming the reason why this was coded up this way was to allow for adding other tags, meaning `docker pull ubuntu:latest` would also add a tag for `ubuntu:14.04` `ubuntu:14.04.1` and `ubuntu:trusty`.

**Comment by dmcgowan** [comment]
BTW I set as code review because I think the behavior for v1 and v2 should be aligned and since we are not adding features to v1, there is no point in going through the design discussion.

**Comment by dkjer** [comment]
@dmcgowan This was mean purely as an optimization for pulling against repositories that have a large number of tags.  It isn't meant to align behavior between v1 and v2.

>  I am assuming the reason why this was coded up this way was to allow for adding other tags, meaning docker pull ubuntu:latest would also add a tag for ubuntu:14.04 ubuntu:14.04.1 and ubuntu:trusty.

Is that how it works now?  The observed behavior I am seeing (in master) is that we only pull a single tag in that case:

```
$ DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.7.0-dev/binary/docker pull ubuntu:latest
latest: Pulling from ubuntu
e118faab2e16: Pull complete 
7e2c5c55ef2c: Pull complete 
e04c66a223c4: Pull complete 
fa81ed084842: Already exists 
ubuntu:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.
Digest: sha256:738edd684282277c07f23277718e43562daf2ee210f7aca9a13fae65f0159ddd
Status: Downloaded newer image for ubuntu:latest

$ DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.7.0-dev/binary/docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
ubuntu              latest              fa81ed084842        10 days ago         188.3 MB
```

I see the same behavior using a v1 registry:
```
$ DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.7.0-dev/binary/docker pull registry.docker/debian:latest
Pulling repository registry.docker/debian
d296866df816: Download complete 
b1159f7d5f0c: Download complete 
Status: Downloaded newer image for registry.docker/debian:latest

$ DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.7.0-dev/binary/docker images
REPOSITORY               TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
registry.docker/debian   latest              d296866df816        3 months ago        132.3 MB
```

I believe this section from TagStore.pullRepository will only tag a single tagsList response:
```
for tag, id := range tagsList {
    if askedTag != "" && tag != askedTag {
        continue
    }
    if err := s.Tag(repoInfo.LocalName, tag, id, true); err != nil {
        return err
    }
}
```

Which means even if I am pulling down thousands of tags, all of them are ignored anyways, except for the one I asked for.  Might as well just request that specific tag, instead?

**Comment by dkjer** [comment]
@dmcgowan It looks like this commit changed the behavior you mentioned:

https://github.com/docker/docker/commit/72a9000fcfa2ec5a2c4a29fb62a17c34e6dd186f

**Comment by thaJeztah** [comment]
Correct, that behaviour was reverted because 
a) it could result in performance issues with a large number of tags (apparently, that wasn't solved) and
b) the same behaviour would not be supported (or could not guaranteed to be supported) by registry V2

**Comment by dmcgowan** [comment]
LGTM

ping @stevvooe @RichardScothern

**Comment by dmcgowan** [comment]
>Conflicts:
	graph/pull.go
	registry/registry_test.go
	registry/session.go

@dkjer from a rebase? can you clean this up

**Comment by RichardScothern** [comment]
LGTM

**Comment by dkjer** [comment]
@dmcgowan Commit message cleaned up

File: graph/pull.go [hunk 1]
```diff
@@ -219,8 +219,11 @@ func (s *TagStore) pullRepository(r *registry.Session, out io.Writer, repoInfo *
 	}
 
 	logrus.Debugf("Retrieving the tag list")
-	tagsList, err := r.GetRemoteTags(repoData.Endpoints, repoInfo.RemoteName)
+	tagsList, err := r.GetRemoteTags(repoData.Endpoints, repoInfo.RemoteName, askedTag)
 	if err != nil {
+		if strings.Contains(err.Error(), "not found") && askedTag != "" {
```
Comment (stevvooe) [inline graph/pull.go]: Could we add a type for this, while we're here? String matching will never work out the way you want.

File: registry/session.go [hunk 2]
```diff
@@ -303,8 +310,19 @@ func (r *Session) GetRemoteTags(registries []string, repository string) (map[str
 		}
 
 		result := make(map[string]string)
-		if err := json.NewDecoder(res.Body).Decode(&result); err != nil {
-			return nil, err
+		if askedTag == "" {
```
Comment (stevvooe) [inline registry/session.go]: Please add a separate method for `GetRemoteTags` and `GetRemoteTag`. These have fundamentally different behaviors and there is no reason to have shared functions.

**Comment by stevvooe** [comment]
The design seems okay, but pushing the switching functionality down to the session is coupling the UI with the data layer. It also makes `GetRemoteTags` more complex. There should be a separate method, `GetRemoteTag` that returns whatever is expected (perhaps, this only returns the image id, rather than a map).

This PR needs to be reworked before proceeding.

**Comment by dkjer** [comment]
@stevvooe PR updated with your suggestions

**Comment by stevvooe** [comment]
@dkjer Thanks!

LGTM

**Comment by estesp** [comment]
ignoring experimental CI failure as it is the docker-py test flakiness and this PR doesn't touch experimental.

LGTM

**Comment by jvgogh** [comment]
@jfrazelle @stevvooe can you confirm this change won't be cherrypicked in to 1.7? The reason I'm asking is that this change is a breaking change for us that will take @mattmoor a week or two to fix on our end

**Comment by jfrazelle** [comment]
no it will not be sorry it is too late now

**Comment by jfrazelle** [comment]
i hope u can understand we CANNOT cherrypick things into the release the day of the release...

**Comment by jvgogh** [comment]
@jfrazelle I think you misunderstood, I wanted to make sure this change WAS NOT cherrypicked last minute, we're on the same page :)

**Comment by mattmoor** [comment]
+1, we just want to make sure we don't have to cherrypick something last minute as well.

thanks.

**Comment by mattmoor** [comment]
@jfrazelle @stevvooe Suggest reverting this, it breaks v1 private image support, including DockerHub.

This makes it such that the first request is:
   /v1/repositories/<name>/tags/<tag>

With subsequent requests to:
  /v1/images/<top-layer>/ancestry

However, even if X-Docker-Token is returned from the first endpoint, it is not captured and passed to subsequent requests (as it should be).

I have confirmed that I CANNOT pull a private v1 image from DockerHub with a client built at this change.
```
$ ~/docker-1.7.0-dev  pull mattmoor/private-test:baz
Pulling repository mattmoor/private-test
5cf8515f9d13: Pulling image (baz) from mattmoor/private-test, endpoint: https://registry-1.docker.io/v5cf8515f9d13: Error pulling image (baz) from mattmoor/private-test, endpoint: https://registry-1.docker.io/v1/, Server error: 400 trying to fetch remote history for 5cf8515f9d13ae92d47ec13b78a031457ff24d75cf8515f9d13: Error pulling image (baz) from mattmoor/private-test, Server error: 400 trying to fetch remote history for 5cf8515f9d13ae92d47ec13b78a031457ff24d77885b22d958114689e315576c 
Error pulling image (baz) from mattmoor/private-test, Server error: 400 trying to fetch remote history for 5cf8515f9d13ae92d47ec13b78a031457ff24d77885b22d958114689e315576c
```

**Comment by mattmoor** [comment]
@dmcgowan see above

**Comment by mattmoor** [comment]
It looks like `/v1/repos/<name>/images` is actually still the first request, but the token isn't passed to `/v1/repos/<name>/tags/<tag>`, or subsequent `/v1/images/...` requests.

FWIW, see: https://gist.github.com/mmdriley/0ed2129a93fd0c15ff36#life-of-a-docker-pull for the previous behavior in detail.

**Comment by mattmoor** [comment]
Sorry for the noise, the Token breakage is an even earlier change, so I'll root cause and send a PR when I've identified the issue.

It is sufficient for us that this not land in 1.7, the new endpoint is trivial enough to implement.


tl;dr The problem manifests when you respond to v2 Ping, but fallback on v1 which breaks Token auth.

This second issue is likely the source of the aforementioned DockerHub breakage as well, but I still need to confirm that with other builds.

Someone may want to check the final 1.7 build, as this smells related to a prior issue we fixed with sending the registry Auth headers.

**Comment by jgkamat** [comment]
+1 on reverting this
I'm using a private docker v1 registry hosted on artifactory, and this change breaks all pulls! If this could be reverted until this issue is fixed, then that would be greatly appreciated! I've tried out reverting this myself and recompiling on current master, and everything goes back to normal.
 
```
root@mechjeb-1:~/docker/singlehost# docker pull docker-registry.company.com:80/test_eng/ubuntu12.04_cluster_tools
Pulling repository docker-registry.company.com:80/test_eng/ubuntu12.04_cluster_tools
Could not reach any registry endpoint
```
Daemon
```
[jay@laythe binary]$ sudo ./docker -d --insecure-registry=docker-registry.company.com:80
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
INFO[0000] [graphdriver] using prior storage driver "aufs" 
WARN[0000] Running modprobe bridge nf_nat failed with message: , error: exit status 1 
WARN[0000] Your kernel does not support cgroup memory limit: mountpoint for memory not found 
WARN[0000] Your kernel does not support cgroup cfs period 
WARN[0000] Your kernel does not support cgroup cfs quotas 
INFO[0000] Loading containers: start.                   
........
INFO[0000] Loading containers: done.                    
INFO[0000] Daemon has completed initialization          
INFO[0000] Docker daemon                                 commit=b349a74 execdriver=native-0.2 graphdriver=aufs version=1.7.0-dev
INFO[0009] POST /v1.20/images/create?fromImage=docker-registry.company.com%3A80%2Ftest_eng%2Fubuntu12.04_cluster_tools%3Alatest 
ERRO[0010] unable to get remote tags: Could not reach any registry endpoint 
```

**Comment by mattmoor** [comment]
@stevvooe As I mentioned to you in person at DockerCon, if I curl this endpoint on DockerHub, I get a malformed response for this:
```
$ curl https://index.docker.io/v1/repositories/library/ubuntu/tags/14.04
[{"pk": 19804636, "id": "6d494699"}, {"pk": 19843084, "id": "9fd3c8c9"}, {"pk": 19804634, "id": "43505007"}, {"pk": 19804633, "id": "428b411c"}]
```

By contrast (this is the correct response, per docker/docker-registry):
```
$ curl https://quay.io/v1/repositories/coreos/etcd/tags/latest
"c2ebda3d7dd843f4eecc59da047674a540c97df9e05c025b029498828bbb23c9"
```

FWIW, I get similar malformed data for repos I know to not have v2 data (e.g. mattmoor/debian:wheezy).

**Comment by stevvooe** [comment]
@mattmoor For hub issues, please report to support.docker.com. I have no clue why this would change. That doesn't look like any response I've seen before, so there may be a problem. (I am trying to track someone down for this...). @bcwalrus @kencochrane 

@mattmoor @jgkamat If you want this PR to be reverted, could you file an issue explaining why? There seem to be a few different problems happening here. It is also possible that this PR exposed an existing auth issue.

@jgkamat What build/version are you running? You may be seeing a hub issue.

**Comment by mattmoor** [comment]
@stevvooe I don't think this should be a problem for us, will confirm tomorrow.  The bigger issue was the other Token issue, which was already merged for 1.7.1

I suspect that @jgkamat is using a nightly or dev build with Artifactory's Docker support.  I'll reach out to the jFrog folks to make sure they are aware of this breakage in Artifactory's registry support, so they can prepare a fix and/or chime in here.


re: DockerHub, I have no clue why it would exhibit this behavior, but I too have never seen it :)

**Comment by jgkamat** [comment]
@stevvooe I am using a dev client of docker, but i'm not sure about the commit version of the docker registry. I git-bisected to this commit, and thats why I'm here...
Thanks @mattmoor! :smile:

**Comment by mattmoor** [comment]
@jfrog Haven't heard back on my email, so I'll try this :)

**Comment by dmcgowan** [comment]
@mattmoor the curl is hitting the index while the Docker daemon will hit the registry. The registry is returning the expected the content. I have no idea what the index is returning. That would be a question for the hub team.

@jgkamat  I am -1 on reverting, the issue with artifactory is that it is not implementing a documented endpoint in the V1 API. While in many cases a subset is possible, there is enough time to bring the implementation up to spec before the next release. Thank you @mattmoor for making them aware of this issue.

**Comment by yoav** [comment]
@jgkamat What version of artifactory have you tested this against? The get tag API should be supported since artifactory 3.8.0+.
We'll double check this, but we don't expect any issues with the new client.

**Comment by jgkamat** [comment]
@yoav, we are using artifatory 3.7, if we experience issues on 3.8, then I'll let you know!

**Comment by stevvooe** [comment]
@yoav @jgkamat I recommend that we take this conversation elsewhere. This is a merged pull request. Please open another issue to discuss or use the mailing list.
